File: it/versions/structures/default/js/foo/v4.js
import * as bar from "bar/v1.js";
import * as bar2 from "bar/v2_0.js";
import * as bar21 from "bar/v2_1.js";

export class Thing {
  constructor(name, other, other2, other21) {
    this.name = name;
    this.other = other;
    this.other2 = other2;
    this.other21 = other21;
  }

  static decode(data) {
    let v_name = data["name"];

    if (v_name !== null && v_name !== undefined) {
      v_name = v_name;
    } else {
      v_name = null;
    }

    let v_other = data["other"];

    if (v_other !== null && v_other !== undefined) {
      v_other = bar.Other.decode(v_other);
    } else {
      v_other = null;
    }

    let v_other2 = data["other2"];

    if (v_other2 !== null && v_other2 !== undefined) {
      v_other2 = bar2.Other.decode(v_other2);
    } else {
      v_other2 = null;
    }

    let v_other21 = data["other21"];

    if (v_other21 !== null && v_other21 !== undefined) {
      v_other21 = bar21.Other.decode(v_other21);
    } else {
      v_other21 = null;
    }

    return new Thing(v_name, v_other, v_other2, v_other21);
  }

  encode() {
    const data = {};

    if (this.name !== null && this.name !== undefined) {
      data["name"] = this.name;
    }

    if (this.other !== null && this.other !== undefined) {
      data["other"] = this.other.encode();
    }

    if (this.other2 !== null && this.other2 !== undefined) {
      data["other2"] = this.other2.encode();
    }

    if (this.other21 !== null && this.other21 !== undefined) {
      data["other21"] = this.other21.encode();
    }

    return data;
  }
}

File: it/versions/structures/default/python/foo/v4.py
import bar.v1 as bar
import bar.v2_0 as bar2
import bar.v2_1 as bar21

class Thing:
  def __init__(self, _name, _other, _other2, _other21):
    self._name = _name
    self._other = _other
    self._other2 = _other2
    self._other21 = _other21

  @property
  def name(self):
    return self._name

  @property
  def other(self):
    return self._other

  @property
  def other2(self):
    return self._other2

  @property
  def other21(self):
    return self._other21

  @staticmethod
  def decode(data):
    f_name = None

    if "name" in data:
      f_name = data["name"]

      if f_name is not None:
        if not isinstance(f_name, unicode):
          raise Exception("not a string")

    f_other = None

    if "other" in data:
      f_other = data["other"]

      if f_other is not None:
        f_other = bar.Other.decode(f_other)

    f_other2 = None

    if "other2" in data:
      f_other2 = data["other2"]

      if f_other2 is not None:
        f_other2 = bar2.Other.decode(f_other2)

    f_other21 = None

    if "other21" in data:
      f_other21 = data["other21"]

      if f_other21 is not None:
        f_other21 = bar21.Other.decode(f_other21)

    return Thing(f_name, f_other, f_other2, f_other21)

  def encode(self):
    data = dict()

    if self._name is not None:
      data["name"] = self._name

    if self._other is not None:
      data["other"] = self._other.encode()

    if self._other2 is not None:
      data["other2"] = self._other2.encode()

    if self._other21 is not None:
      data["other21"] = self._other21.encode()

    return data

  def __repr__(self):
    return "<Thing name:{!r}, other:{!r}, other2:{!r}, other21:{!r}>".format(self._name, self._other, self._other2, self._other21)
